backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        public IEnumerable<ProjectSnapshot> Refresh(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+            var report = Get(reportId);
+            var currentSnapshots = this.Context.ProjectSnapshots
+                .Include(s => s.Project).ThenInclude(p => p.Agency)
+                .Include(s => s.Project).ThenInclude(p => p.Status)
+                .Where(s => s.SnapshotOn == report.To)
+                .ToArray();
+
+            var snapshots = GenerateSnapshots(DateTime.UtcNow).Concat(currentSnapshots);
+            foreach (ProjectSnapshot snapshot in snapshots)
+            {
+                var previousSnapshot = snapshots.Where(s => s.ProjectId == snapshot.ProjectId && s.Id != snapshot.Id).FirstOrDefault();
+                snapshot.BaselineIntegrity = (snapshot?.NetProceeds ?? 0) - (previousSnapshot?.NetProceeds ?? 0);
+            }
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Remove the specified report from the datasource.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project report rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to delete report.</exception>
+        /// <returns></returns>
+        public void Remove(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var originalReport = this.Context.ProjectReports
+                .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
+            if (originalReport.IsFinal) throw new InvalidOperationException($"Unable to delete FINAL project reports.");
+
+            this.Context.ProjectReports.Remove(originalReport);
+            this.Context.CommitTransaction();
+        }
+
+
+        /// <summary>
+        /// Generate snapshots for all SPL projects using the passed 'to' date as the snapshot date.
+        /// </summary>
+        /// <param name="to"></param>
+        /// <returns></returns>
+        private IEnumerable<ProjectSnapshot> GenerateSnapshots(DateTime to)
+        {
+            var splProjects = this.Context.Projects
+                .Include(p => p.Agency)
+                .Include(p => p.Status)
+                .Where(p => p.Workflow.Code == "SPL") ?? throw new KeyNotFoundException();
+
+            IEnumerable<ProjectSnapshot> projectSnapshots = new ProjectSnapshot[0];
+            foreach (Project project in splProjects)
+            {
+                this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+                ProjectSnapshot snapshot = new ProjectSnapshot(project);
+                snapshot.SnapshotOn = to;
```

Comment:
All the snapshots should have the same `to` value, but it doesn't need to be passed into the function.  It can be generated when the function is called.

Reply:
So here was my reasoning for this:
Some of the callers of this function are actually saving a report to the DB. In order for the snapshots generated for that report to actually be connected, the To value needs to be identical (Note that this function is a private, internal helper function, not a service function).

I think every time you call DateTime.UtcNow you will get a new, slightly progressed value. This would therefore be a problem if the report was saved with an earlier To: value using DateTime.UtcNow, and then updated with a slightly progressed To: value obtained by calling DateTime.UtcNow again.

Reply:
You wouldn't call `DateTime.UtcNow` multiple times.  Just once at the beginning of the function.  Having it as a parameter implies the caller decides.  Which would mean that the data should reflect that specified `to` date and time.  This function doesn't do that.  We grab the most recent project information and generate new snapshots.

Reply:
I think what you are asking for is for GenerateSnapshots to be turned into this:

```
private IEnumerable<ProjectSnapshot> GenerateSnapshots()
        {
            var to = DateTime.UtcNow;
            var splProjects = this.Context.Projects
                .Include(p => p.Agency)
                .Include(p => p.Status)
                .Where(p => p.Workflow.Code == "SPL") ?? throw new KeyNotFoundException();

            IEnumerable<ProjectSnapshot> projectSnapshots = new ProjectSnapshot[0];
            foreach (Project project in splProjects)
            {
                this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);

                ProjectSnapshot snapshot = new ProjectSnapshot(project);
                snapshot.SnapshotOn = to;

                projectSnapshots = projectSnapshots.Append(snapshot);
            }
            return projectSnapshots;
        }
```
If the above is what you were thinking, I debugged this and it does not work.
Here is an example:
- The API calls the Add function. The add function saves the current DateTime.UtcNow to a local variable. (debugged value: 23:27:42.5129541)
- The add function creates a new report using the above DateTime.UtcNow() value.
- The add function calls the above GenerateSnapshots function. 
- The GenerateSnapshots saves another, slightly progressed value of DateTime.UtcNow (debugged value: 23:27:44.6235520)
- The GenerateSnapshots function generates all of the snapshots using the same DateTime.UtcNow value from the previous step.
- Now we have a report with a To value of 23:27:42.5129541 and a whole bunch of snapshots with a snapshotOn value of 23:27:44.6235520. Since the To value doesn't match the snapshotOn value, no data is returned if this report is requested.

Now interestingly, when I tested this without debugging I had this pass a few times. Apparently this is because the UtcNow function is only accurate to a few ms. As a result its possible that two calls happening in quick succession could return the same ms value. However, I wouldn't want to depend on that.

Reply:
The date used needs to be identical everywhere.  If the `GenerateSnapshots()` is setting the datetime then it needs to be used by the calling code.  This is where it's a little ugly and I don't like my simple solution - The calling code could retrieve the datetime from the first snapshot returned.  This highlights something not great about the implementation I'm suggesting though.  However if you pass the date, then in theory it should be used as part of the query (although technically we can do whatever we want).